include/zephyr/drivers/sip_svc/sip_svc_driver.h

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2023, Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SIP_SVC_DRIVER_
+#define ZEPHYR_INCLUDE_SIP_SVC_DRIVER_
+
+#include <zephyr/device.h>
+#include <zephyr/arch/arm64/arm-smccc.h>
+#include <zephyr/drivers/sip_svc/sip_svc_proto.h>
+
+#define SVC_NAME_LENGTH 4
+
+struct svc_data {
+	char *method;
+	void (*svc_fn)(unsigned long function_id, unsigned long arg0,
+					 unsigned long arg1, unsigned long arg2,
+					 unsigned long arg3, unsigned long arg4,
+					 unsigned long arg5, unsigned long arg6,
+					 struct arm_smccc_res *res);
+};
+
+struct svc_api {
+	void (*svc_func)(const struct device *dev, unsigned long function_id,
+					 unsigned long arg0, unsigned long arg1, unsigned long arg2,
+					 unsigned long arg3, unsigned long arg4, unsigned long arg5,
+					 unsigned long arg6, struct arm_smccc_res *res);
+	bool (*sip_svc_plat_func_id_valid)(uint32_t command, uint32_t func_id);
+	uint32_t (*sip_svc_plat_format_trans_id)(uint32_t client_idx,
+					 uint32_t trans_idx);
+	uint32_t (*sip_svc_plat_get_trans_idx)(uint32_t trans_id);
+	void (*sip_svc_plat_update_trans_id)(struct sip_svc_request *request,
+					 uint32_t trans_id);
+	void (*sip_svc_plat_free_async_memory)(struct sip_svc_request *request);
+	int (*sip_svc_plat_async_res_req)(unsigned long *a0, unsigned long *a1,
+					 unsigned long *a2, unsigned long *a3, unsigned long *a4,
+					 unsigned long *a5, unsigned long *a6, unsigned long *a7,
+					 char *buf, size_t size);
+	int (*sip_svc_plat_async_res_res)(struct arm_smccc_res *res, char *buf,
+					 size_t *size, uint32_t *trans_id);
+	uint32_t (*sip_svc_plat_get_error_code)(struct arm_smccc_res *res);
+};
```

Comment:
This is a really odd driver that is basically not following any of the standard (zephyr) way of defining and using drivers. There is a reason for that?

Take for example (this is valid for any zephyr driver subsystem really) the CAN driver subsystem. You have a `__subsystem` marked set of hooks in `include/zephyr/drivers/can.h` that are specialized by a driver without resorting to `__weak` symbols like it is done in your implementation (and without any "default / empty" implementation for that).

Reply:
thanks for pointing this out, I will rework the current implementation to this

Reply:
I have reworked the implementation with the suggested changes. Could you please check them?